nengo/params.py

```diff
@@ -416,6 +426,47 @@ def validate(self, instance, function_info):
         super(FunctionParam, self).validate(instance, function)
 
 
+class CopyableObject(object):
+    """Mixin to allow to create copies of instances of classes with parameters.
+
+    You can add a list of strings as class attribute _param_init_order to
+    declare the order in which parameters have to be initialized. Missing
+    parameters will be initialized last in an undefined order.
+    """
+
+    def __getstate__(self):
+        sp = super(CopyableObject, self)
+        if hasattr(sp, '__getstate__'):
+            state = sp.__getstate__()
+        else:
+            state = dict(self.__dict__)
+
+        for attr in params(self):
+            param = getattr(self.__class__, attr)
+            if self in param:
+                state[attr] = getattr(self, attr)
+
+        return state
+
+    def __setstate__(self, state):
+        if hasattr(self.__class__, '_param_init_order'):
```

Comment:
Not a big deal, but can we define this in `CopyableObject` so that we don't have to do the `hasattr` here?